src/ai/backend/storage/kmanila/__init__.py

```diff
@@ -0,0 +1,272 @@
+            },
+        ) as resp:
+            result = await resp.json()
+            access_list: list[dict[str, Any]] = result["access_list"]
+            log_items = [
+                str({"access_level": item["access_level"], "access_to": item["access_to"]})
+                for item in access_list
+            ]
+            log.debug(f"Found {len(access_list)} access control items. ({', '.join(log_items)})")
+            if access_list:
+                return False
+
+        request_data = {
+            "os-allow_access": {
+                "access_level": self.access_level,
+                "access_type": "ip",
+                "access_to": self.access_to,
+            }
+        }
+        async with session.post(
+            f"/d3/adm/manila/v2/{project_id}/shares/{volume_id}/action",
+            json=request_data,
+        ) as resp:
+            return True
+
+    async def _mount_volumes_to_agents(self) -> None:
+        # Do mount request to GPU nodes through ssh
+        def _build_ssh_cmd(agent_addr: str) -> list[str]:
+            return ["ssh", agent_addr]
+
+        for agent_addr in self.agent_addrs:
+            cmds = _build_ssh_cmd(agent_addr)
+            await asyncio.create_subprocess_exec(*cmds)
+
+    async def create_quota_scope(
+        self,
+        quota_scope_id: QuotaScopeID,
+        options: QuotaConfig | None = None,
+        extra_args: dict[str, Any] | None = None,
+    ) -> None:
+        volume_info: dict[str, str] = kmanila_request_params.check(extra_args)
+        volume_name = volume_info["volume_name"]
+
+        auth_info = await self._get_auth_info(quota_scope_id)
+        headers = {
+            "X-Auth-Token": auth_info["auth_token"],
+            "Content-Type": "application/json",
+        }
+        async with aiohttp.ClientSession(base_url=self.api_base_url, headers=headers) as sess:
+            volume_id = await self._create_volume(
+                sess, quota_scope_id, options, auth_info, name=volume_name
+            )
+            await self._create_access_control(sess, quota_scope_id, volume_id, auth_info)
+
+
+class KManilaFSVolume(BaseVolume):
+    async def create_quota_model(self) -> AbstractQuotaModel:
+        access_level = self.config.get("access_level", "rw")
+        max_poll_count = (
+            int(self.config.get("max_poll_count"))  # type: ignore[arg-type]
+            if self.config.get("max_poll_count")
+            else DEFAULT_MAX_POLL_COUNT
+        )
+        return KManilaQuotaModel(
+            self.mount_path,
+            self.etcd,
+            URL(self.config["api_base_url"]),
+            self.config["agent_addrs"],
+            access_to=self.config["access_to"],
+            access_level=access_level,
+            kmanila_requestor_id=self.config["user_id"],
+            kmanila_requestor_pwd=self.config["user_password"],
+            default_project_id=self.config.get("project_id"),
+            default_netword_id=self.config.get("netword_id"),
```

Comment:
Just to double check, is `netword` a typo of `network` or is it intended?

Reply:
wow what a huge mistypo. I fixed it!